src/Analyzers/CSharp/Analyzers/ConstantInterpolatedString/CSharpConstantInterpolatedStringAnalyzer.cs

```diff
@@ -0,0 +1,82 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using Microsoft.CodeAnalysis.CodeStyle;
+using Microsoft.CodeAnalysis.CSharp.Shared.Extensions;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Operations;
+
+namespace Microsoft.CodeAnalysis.CSharp.ConstantInterpolatedString
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)] // constant interpolated strings is a C#-only feature.
+    internal class CSharpConstantInterpolatedStringAnalyzer : AbstractBuiltInCodeStyleDiagnosticAnalyzer
+    {
+        public CSharpConstantInterpolatedStringAnalyzer() : base(
+            IDEDiagnosticIds.UseConstantInterpolatedStringDiagnosticId,
+            EnforceOnBuildValues.AddAccessibilityModifiers,
+            option: null,
+            new LocalizableResourceString(nameof(CSharpAnalyzersResources.Use_constant_interpolated_string), CSharpAnalyzersResources.ResourceManager, typeof(CSharpAnalyzersResources)))
+        {
+        }
+
+        public override DiagnosticAnalyzerCategory GetAnalyzerCategory() => DiagnosticAnalyzerCategory.SemanticSpanAnalysis;
+
+        protected override void InitializeWorker(AnalysisContext context)
+        {
+            context.RegisterCompilationStartAction(context =>
+            {
+                if (!((CSharpCompilation)context.Compilation).LanguageVersion.HasConstantInterpolatedStrings())
+                {
+                    return;
+                }
+
+                context.RegisterOperationAction(context =>
+                {
+                    var operation = (IBinaryOperation)context.Operation;
+                    var info = operation.SemanticModel!.GetSymbolInfo(operation.Syntax, context.CancellationToken);
+                    if (!ShouldAnalyze(operation))
+                    {
+                        return;
+                    }
+
+                    if (!AllOperandsAreStringLiterals(operation))
+                    {
+                        context.ReportDiagnostic(Diagnostic.Create(Descriptor, operation.Syntax.GetLocation()));
+                    }
+                }, OperationKind.Binary);
+            });
+        }
+
+        private static bool AllOperandsAreStringLiterals(IBinaryOperation? operation)
+        {
+            if (operation is null)
+            {
+                return false;
+            }
+
+            return operation.LeftOperand.Kind == OperationKind.Literal &&
+                (operation.RightOperand.Kind == OperationKind.Literal || AllOperandsAreStringLiterals(operation.RightOperand as IBinaryOperation));
```

Comment:
😕 Should this check the type as well?

Reply:
@sharwell This is done in `ShouldAnalyze` by the `operation.ConstantValue.Value is not string` check